File: backend/app/routers/stops.py
from typing import Optional
from fastapi import APIRouter, Query, HTTPException
from app.schemas import (
    GeoJsonFeatureCollection,
    StopProfileResponse,
    StopRankingResponse,
    StopBatchRequest,
    StopBatchResponse,
)
from app import spatial_engine

router = APIRouter(tags=["Physical Stops (Micro)"])


@router.get("/stops", response_model=GeoJsonFeatureCollection)
async def get_stops(city: str = Query(..., description="City slug (e.g. kielce, warszawa, gzm)")):
    """Returns GeoJSON FeatureCollection of physical transit stops from stop_dna.gpkg with micro & macro Stop DNA metrics."""
    try:
        return spatial_engine.get_stops(city)
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Spatial error: {str(e)}")


@router.get("/stops/ranking", response_model=StopRankingResponse)
async def get_stops_ranking(
    city: str = Query(..., description="City slug"),
    order_by: str = Query("stop_local_score_raw", description="Metric column name to sort by"),
    order_dir: str = Query("desc", description="Sort direction ('asc' or 'desc')"),
    limit: int = Query(20, ge=1, le=1000, description="Max number of items to return"),
    offset: int = Query(0, ge=0, description="Offset for pagination"),
    rank: Optional[int] = Query(None, ge=1, description="1-based exact position to fetch"),
    grade: Optional[str] = Query(None, description="Filter by grade (e.g. A+, A, B, C, D, E, F)"),
    is_hub_anchor: Optional[bool] = Query(None, description="Filter for hub anchor stops"),
    min_departures: Optional[float] = Query(None, description="Filter for minimum departures per hour"),
    min_pop: Optional[float] = Query(None, description="Filter for minimum population"),
    h3_index: Optional[str] = Query(None, description="Filter for H3 cell index"),
):
    """Universal ranking endpoint for physical stops with arbitrary limits and dynamic sorting."""
    try:
        return spatial_engine.get_stops_ranking(
            city=city,
            order_by=order_by,
            order_dir=order_dir,
            limit=limit,
            offset=offset,
            rank=rank,
            grade=grade,
            is_hub_anchor=is_hub_anchor,
            min_departures=min_departures,
            min_pop=min_pop,
            h3_index=h3_index,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ranking error: {str(e)}")


@router.post("/stops/batch", response_model=StopBatchResponse)
async def get_stops_batch(payload: StopBatchRequest):
    """Batch lookup endpoint retrieving full profiles for multiple stop IDs in a single query."""
    try:
        return spatial_engine.get_stops_batch(city=payload.city, stop_ids=payload.stop_ids)
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Batch error: {str(e)}")


@router.get("/stops/{stop_id}", response_model=StopProfileResponse)
async def get_stop_profile(
    stop_id: str,
    city: str = Query(..., description="City slug")
):
    """Returns deep Stop DNA profile of a single physical stop from stop_dna.gpkg."""
    try:
        return spatial_engine.get_stop_profile(city=city, stop_id=stop_id)
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Stop profile error: {str(e)}")

File: urban-dashboard/src/lib/api/stops.ts
import { apiFetch } from "./client";
import type {
  GeoJsonFeatureCollection,
  StopRankingResponse,
  StopRankingItem,
  StopBatchResponse,
  StopProfileResponse,
} from "./types";

export interface StopRankingParams {
  city: string;
  order_by?: string;
  order_dir?: "asc" | "desc";
  limit?: number;
  offset?: number;
  rank?: number;
  grade?: string;
  is_hub_anchor?: boolean;
  min_departures?: number;
  min_pop?: number;
  h3_index?: string;
  query?: string;
}

export async function fetchStopsGeoJson(
  city: string,
  signal?: AbortSignal
): Promise<GeoJsonFeatureCollection> {
  try {
    return await apiFetch<GeoJsonFeatureCollection>(
      `/api/v1/stops?city=${encodeURIComponent(city)}`,
      { signal }
    );
  } catch (e: any) {
    if (e?.name === "AbortError") throw e;
    const local = await fetch(`/data/showcase/${city}/stops.json`, { signal }).catch(() => null);
    if (local && local.ok) return await local.json();
    return { type: "FeatureCollection", features: [] };
  }
}

export async function fetchStopsRanking(
  params: StopRankingParams,
  signal?: AbortSignal
): Promise<StopRankingResponse> {
  const q = new URLSearchParams({ city: params.city });
  if (params.order_by) q.set("order_by", params.order_by);
  if (params.order_dir) q.set("order_dir", params.order_dir);
  if (params.limit != null) q.set("limit", String(params.limit));
  if (params.offset != null) q.set("offset", String(params.offset));
  if (params.rank != null) q.set("rank", String(params.rank));
  if (params.grade) q.set("grade", params.grade);
  if (params.is_hub_anchor != null) q.set("is_hub_anchor", String(params.is_hub_anchor));
  if (params.min_departures != null) q.set("min_departures", String(params.min_departures));
  if (params.min_pop != null) q.set("min_pop", String(params.min_pop));
  if (params.h3_index) q.set("h3_index", params.h3_index);
  if (params.query) q.set("query", params.query);

  return apiFetch<StopRankingResponse>(`/api/v1/stops/ranking?${q.toString()}`, { signal });
}

export async function searchStops(
  city: string,
  query: string,
  limit: number = 20,
  signal?: AbortSignal
): Promise<StopRankingItem[]> {
  const q = new URLSearchParams({ city, query, limit: String(limit) });
  try {
    const res = await apiFetch<StopRankingResponse>(`/api/v1/stops/search?${q.toString()}`, { signal });
    return res.items || [];
  } catch (e: any) {
    if (e?.name === "AbortError") throw e;
    try {
      const res = await apiFetch<StopRankingResponse>(`/api/v1/stops/ranking?${q.toString()}`, { signal });
      return res.items || [];
    } catch {
      return [];
    }
  }
}

export async function fetchStopsBatch(
  city: string,
  stopIds: string[],
  signal?: AbortSignal
): Promise<StopBatchResponse> {
  return apiFetch<StopBatchResponse>("/api/v1/stops/batch", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ city, stop_ids: stopIds }),
    signal,
  });
}

export async function fetchStopProfile(
  city: string,
  stopId: string,
  signal?: AbortSignal
): Promise<StopProfileResponse> {
  return apiFetch<StopProfileResponse>(
    `/api/v1/stops/${encodeURIComponent(stopId)}?city=${encodeURIComponent(city)}`,
    { signal }
  );
}
